Apps/Src/ReadCarCAN.c

```diff
@@ -5,256 +5,482 @@
+                SOC = (*(uint32_t*) &dataBuf.data)/(1000000);  // Convert to integer percent
                 UpdateDisplay_SetSOC(SOC);
                 break;
             }
-            default: 
-                break;
+        default: {  
+            break; // Unhandled CAN message IDs, do nothing
+            }   
         }
     }
 }
+
+/**
+ * @brief error handler callback for disabling charging, 
+ * kills contactor and turns off display
+ */ 
+static void handler_ReadCarCAN_chargeDisable(void) {
+    // Fills buffers with disable messages
+    memset(HVArrayChargeMsgBuffer, DISABLE_SATURATION_MSG, sizeof(HVArrayChargeMsgBuffer));
+
+    // mark regen as disabled and update saturation
+    updateHVArraySaturation(DISABLE_SATURATION_MSG);
+
+    // Kill contactor using a direct write to avoid blocking calls when the scheduler is locked
+    BSP_GPIO_Write_Pin(CONTACTORS_PORT, ARRAY_PRECHARGE_BYPASS_PIN, false);
+
+    // Check that the contactor was successfully turned off
+    bool ret = (bool)Contactors_Get(ARRAY_PRECHARGE_BYPASS_CONTACTOR);
+
+    if(ret != false) { // Contactor failed to turn off; display the evac screen and infinite loop
+        Display_Evac(SOC, SBPV);
+        while(1){;}
+    }
+}
+
+/**
+ * @brief error handler callback for disabling charging, 
+ * kills contactor and turns off display
+ */ 
+static void handler_ReadCarCAN_contactorsDisable(void) {
+
+    // Mark regen as disabled and update saturation
+    updateHVArraySaturation(DISABLE_SATURATION_MSG);
+    updateHVPlusMinusSaturation(DISABLE_SATURATION_MSG);
+
+    // Kill contactor using a direct write to avoid blocking calls when the scheduler is locked
+    BSP_GPIO_Write_Pin(CONTACTORS_PORT, ARRAY_PRECHARGE_BYPASS_PIN, false);
+    BSP_GPIO_Write_Pin(CONTACTORS_PORT, MOTOR_CONTROLLER_PRECHARGE_BYPASS_PIN, false);
+
+    // Fills buffers with disable messages
+    memset(HVArrayChargeMsgBuffer, DISABLE_SATURATION_MSG, sizeof(HVArrayChargeMsgBuffer));
+    memset(HVPlusMinusChargeMsgBuffer, DISABLE_SATURATION_MSG, sizeof(HVPlusMinusChargeMsgBuffer));
+
+    // Updates the saturation with disable
+    updateHVArraySaturation(DISABLE_SATURATION_MSG);
+    updateHVPlusMinusSaturation(DISABLE_SATURATION_MSG);
+
+    // Check that the contactor was successfully turned off
+    bool ret = (bool)Contactors_Get(ARRAY_PRECHARGE_BYPASS_CONTACTOR) || (bool)Contactors_Get(MOTOR_CONTROLLER_PRECHARGE_BYPASS_CONTACTOR);
+
+    if(ret != false) { // Contactor failed to turn off; display the evac screen and infinite loop
+         Display_Evac(SOC, SBPV);
+         while(1){;}
+    }
+}
+
+/**
+ * @brief error handler function to display the evac screen if we get a BPS trip message.
+ * Callbacks happen after displaying the fault, so this screen won't get overwritten
+ */ 
+static void handler_ReadCarCAN_BPSTrip(void) {
+    chargeEnable = false; // Not really necesssary but makes inspection less confusing
+	Display_Evac(SOC, SBPV);   // Display evacuation screen
```

Comment:
Tabbing looks weird on my end. Also 3 s's are in necessary when there should only be 2.